enhancements/microshift/advertise-version-details.md

```diff
@@ -0,0 +1,302 @@
+applying rules based on the values.
+
+### Workflow Description
+
+#### Determining the Version of a Cluster via the API
+
+API clients can read the `openshift/version` ConfigMap contents in the
+normal way.
+
+#### Determining the Version of a Cluster via the CLI
+
+The `oc login` command reports the `message` value of the
+`openshift/motd` ConfigMap when a user successfully authenticates. The
+command should be extended to include the `version` string of the
+`openshift/version` ConfigMap, when it is present.
+
+The `oc version` command reports version details today. Some of the
+information it reports comes from the ClusterVersion resource, which
+is not visible to all users. The command should be updated to look at
+the `openshift/version` ConfigMap when it is present and the
+ClusterVersion resource cannot be read.
+
+#### Publishing the Version in MicroShift
+
+MicroShift will update the `openshift/version` ConfigMap contents on
+startup using a value compiled into the main binary. Upgrading
+MicroShift requires restarting the service, so there is no need to
+continue reconciling the contents of the ConfigMap.
+
+#### Publishing the Version in OCP
+
+OpenShift running in standalone or external control-plane
+configurations may be upgraded dynamically in place without loss of
+API connectivity. The cluster-version-operator will therefore update
+the contents of the ConfigMap as the version of the cluster
+changes. During an upgrade, we want to report the current version as
+the one from which the cluster is upgrading and not expose that an
+upgrade is in progress. The ConfigMap should therefore always contain
+the value of the most recent version marked as `Completed` in the
+history.
+
+### API Extensions
+
+None
+
+### Risks and Mitigations
+
+There is some risk that an API caller could use the version
+information to determine whether the cluster is vulnerable to a known
+exploit. API callers will need to be authenticated to read the data,
+and they could just try the exploit directly without the version
+information.
+
+It is possible on first launch of a cluster for the API server to
+respond before the ConfigMap has been written, and therefore for the
+caller to fail to detect that the product is an OpenShift
+variant. This can be mitigated by populating the data as early as
+possible using an empty ConfigMap with only the `product` field filled
+in.
+
+There is a similar race condition when restarting MicroShift after an
+upgrade when the wrong version information may be presented. Clients
+are not as likely to need to query the MicroShift version at runtime
+in a production setting, because most deployments will either use
+manifests built into the image or use an agent to control application
+deployment from a central server. This race should therefore only
+affect developer scenarios.
+
+### Drawbacks
+
+None
+
+## Design Details
+
+### Open Questions
+
+1. Should we use a different namespace, like `kube-public`?
```

Comment:
We already have `openshift` available for these purposes, see my earlier comment.